lib/sacastats_web/controllers/page_controller.ex

```diff
@@ -1,10 +1,74 @@
 defmodule SacaStatsWeb.PageController do
   use SacaStatsWeb, :controller
 
+  require Logger
+
+  alias SacaStatsWeb.DiscordAuth
+
   def index(conn, _params) do
     render(conn, "index.html")
   end
 
+  def login(conn, _params) do
+    render(conn, "login.html")
+  end
+
+  def login_discord(conn, _params) do
+    client = DiscordAuth.new()
+    auth_url = DiscordAuth.authorize_url!(client)
+
+    conn
+    |> put_session(:client, client)
+    |> redirect(external: auth_url)
+  end
+
+  def login_discord_callback(conn, %{"code" => auth_code, "state" => state}) do
+    IO.inspect get_session(conn, :client), label: "This should be the client from the session"
+    case get_session(conn, :client) do
+      %OAuth2.Client{params: %{state: ^state}} = client ->
+        IO.inspect "state is correct!"
```

Comment:
What is the state?

Reply:
It's a unique key included in the entire OAuth flow [for security purposes](https://discord.com/developers/docs/topics/oauth2#state-and-security). From the docs I understand it's an extra piece of info included in the discourse with the OAuth servers to make sure we're actually talking to Discord.

Reply:
Ah, okay. That makes sense.